fondant/pipeline.py

```diff
@@ -0,0 +1,447 @@
+        self.description = pipeline_description
+        self.package_path = f"{pipeline_name}.tgz"
+        self._graph: t.OrderedDict[str, t.Any] = OrderedDict()
+
+    def add_op(
+        self,
+        task: FondantComponentOp,
+        dependencies: t.Optional[
+            t.Union[FondantComponentOp, t.List[FondantComponentOp]]
+        ] = None,
+    ):
+        """
+        Add a task to the pipeline with an optional dependency.
+
+        Args:
+            task: The task to add to the pipeline.
+            dependencies: Optional task dependencies that needs to be completed before the task
+             can run.
+        """
+        if dependencies is None:
+            dependencies = []
+        elif not isinstance(dependencies, list):
+            dependencies = [dependencies]
+
+        if len(dependencies) > 1:
+            warnings.warn(
+                f"Multiple component dependencies provided for component `{task.name}`. "
+                f"The current version of Fondant can only handle components with a single "
+                f"dependency. Please note that the behavior of the pipeline may be unpredictable"
+                f" or incorrect."
+            )
+        if dependencies:
+            dependencies = [dependency.name for dependency in dependencies]
+
+        task_name = task.name
+        self._graph[task_name] = {
+            "fondant_component_op": task,
+            "dependencies": dependencies,
+        }
+
+    def sort_graph(self):
+        """Sort the graph topologically based on task dependencies."""
+        sorted_graph = []
+        visited = set()
+
+        def depth_first_traversal(node: str):
+            """
+            Perform a depth-first traversal of the graph and its dependencies.
+
+            Args:
+                node: The name of the starting node for traversal.
+            """
+            if node not in visited:
+                visited.add(node)
+                for dependency in self._graph[node]["dependencies"]:
+                    depth_first_traversal(dependency)
+                sorted_graph.append(node)
+
+        for graph_node in self._graph:
+            depth_first_traversal(graph_node)
+
+        self._graph = OrderedDict((node, self._graph[node]) for node in sorted_graph)
+
+    def _validate_pipeline_definition(self, run_id: str):
+        """
+        Validates the pipeline definition by ensuring that the input and output subsets and their
+         associated fields match and are invoked in the correct order.
+
+        Raises:
+            InvalidPipelineDefinition: If a component is trying to invoke a subset that is not
+             defined or created in previous components, or if an invoked subset's schema does not
+              match the previously created subset definition.
+            base_path: the base path where to store the pipelines artifacts
+            run_id: the run id of the component
+        """
+        load_component = True
```

Comment:
This assumes that a pipeline always starts with a load component, which I don't think should be the case. What if you eg. want to run 2 fondant pipelines after each other? But we can keep this assumption for now and tackle it later.

Reply:
yeah indeed, I'v added a todo to tackle this later on but for now it should be fine